### PR TITLE
{deepin/dde}-kwin: init

### PR DESCRIPTION
**Pull request by wineee**
###### Description of changes
https://github.com/linuxdeepin/dde-nixos/issues/9

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by wineee** [comment]
@ofborg build deepin.deepin-kwin

File: pkgs/desktops/deepin/core/deepin-kwin/default.nix [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    xorg.libxcb
+    xorg.libXdmcp
+    xorg.libXcursor
+    xorg.xcbutilcursor
+    xorg.libXtst
+  ];
+
+  cmakeFlags = [
+    "-DKWIN_BUILD_RUNNERS=OFF"
+    "-DDEFINE_DDE_DOCK_PATH=/run/current-system/sw/bin"
```
Comment (NickCao) [inline pkgs/desktops/deepin/core/deepin-kwin/default.nix:108]: What's this for?

Comment (wineee) [reply pkgs/desktops/deepin/core/deepin-kwin/default.nix:108]: https://github.com/linuxdeepin/deepin-kwin/blob/86a4db4cc1cf82c92719c2644086ac4110df4c99/src/effects/screenshot/screenshotdbusinterface1.cpp#L513
Used for blocking the notification when dde-dock calling the screenshot interface，the preview function of  dde-dock will frequently call this interface

Comment (NickCao) [reply pkgs/desktops/deepin/core/deepin-kwin/default.nix:108]: Shall we hardcode the path to dde-dock.

Comment (wineee) [reply pkgs/desktops/deepin/core/deepin-kwin/default.nix:108]: For security reasons, it should not search from `PATH`

Comment (NickCao) [reply pkgs/desktops/deepin/core/deepin-kwin/default.nix:108]: I mean we should just do `-DDEFINE_DDE_DOCK_PATH=${dde-dock}/bin`

Comment (wineee) [reply pkgs/desktops/deepin/core/deepin-kwin/default.nix:108]: It seems that there has a problem in upstream，cmdline for 'dde-dock' is `dde-dock -r`,   not an absolute path(started by startdde)

File: pkgs/desktops/deepin/core/deepin-kwin/default.nix [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    kidletime
+    knotifications
+    kpackage
+    plasma-framework
+    kcmutils
+    knewstuff
+    kdecoration
+    kscreenlocker
+
+    valgrind
```
Comment (SuperSandro2000) [inline pkgs/desktops/deepin/core/deepin-kwin/default.nix:96]: Do we really need valgrind all the time?

File: pkgs/desktops/deepin/core/deepin-kwin/default.nix [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+, lcms2
+, xorg
+}:
+
+stdenv.mkDerivation rec {
+  pname = "deepin-kwin";
+  version = "5.24.3-deepin.1.9";
+
+  /*
+    Here are no buildable tag in github:
```
Comment (SuperSandro2000) [inline pkgs/desktops/deepin/core/deepin-kwin/default.nix:43]:
```suggestion
    There are no buildable tag in github:
```

File: pkgs/desktops/deepin/core/dde-kwin/default.nix [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    repo = pname;
+    rev = "b5c00527b86f773595c786c8015d60f8be3a681b";
+    sha256 = "sha256-qXN9AwjLnqO5BpnrX5PaSCKZ6ff874r08ubCMM272tA=";
+  };
+
+  /*
+    This is the final version of dde-kwin, upstream has been archived.
+    We should remove this package when deepin-kwin release a new version.
+    We need revert Dtk::Core target in postPatch,
+    since the new dtkcore will break other deepin applications.
```
Comment (SuperSandro2000) [inline pkgs/desktops/deepin/core/dde-kwin/default.nix:39]: Does the deepin-kwin packaged below not work?

Comment (wineee) [inline pkgs/desktops/deepin/core/dde-kwin/default.nix:39]: dde-kwin is not deepin-kwin，but cmake break some deepin applications，maybe it can be solved

**Review by wegank — APPROVED** [review]

File: pkgs/desktops/deepin/core/dde-kwin/default.nix [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    We should remove this package when deepin-kwin release a new version.
+  */
+
+  postPatch = ''
+    substituteInPlace CMakeLists.txt \
+      --replace "/usr/include/KWaylandServer" "${kwayland.dev}/include/KWaylandServer"
+    substituteInPlace deepin-wm-dbus/deepinwmfaker.cpp \
+      --replace "/usr/lib/deepin-daemon" "/run/current-system/sw/lib/deepin-daemon" \
+      --replace "/usr/share/backgrounds" "/run/current-system/sw/share/backgrounds" \
+      --replace "/usr/share/wallpapers" "/run/current-system/sw/share/wallpapers"
```
Comment (NickCao) [inline pkgs/desktops/deepin/core/dde-kwin/default.nix:46]: Still not happy with this part, but we don't have better alternatives now anyway.

**Pull request merged**